### ISSUE TITLE
Why Union type is not following "one of provided types" pattern

### ISSUE THREAD

**Issue by kresli** [Microsoft/TypeScript#30614]
```typescript
interface IA {
  a: string;
}

interface IB {
  b: string
  c: string
}

type IInput = IB | IA;

const a: IInput = { a: '', b: '' } // WRONG Should Error
const b: IInput = { a: '', c: '' } // WRONG Should Error
const c: IInput = { a: '', b: '', c: '' } // WRONG Should Error
const d: IInput = { a: '' } // GOOD PASS
const e: IInput = { b: '' } // GOOD ERROR
const f: IInput = { b: '', c: ''  } // GOOD PASS
```

**Comment by dragomirtitian** [comment]
This is the same issue as this: #20863. I also have a decent writeup of it [here](https://stackoverflow.com/questions/52677576/typescript-discriminated-union-allows-invalid-state/52678379#52678379), and a solution in the for of `StrictUnion`: 

```ts
type UnionKeys<T> = T extends any ? keyof T : never;
type StrictUnionHelper<T, TAll> = T extends any ? T & Partial<Record<Exclude<UnionKeys<TAll>, keyof T>, never>> : never;
type StrictUnion<T> = StrictUnionHelper<T, T>

interface IA {
  a: string;
}

interface IB {
  b: string
  c: string
}

type IInput = StrictUnion<IB | IA>;

const a: IInput = { a: '', b: '' } // err
const b: IInput = { a: '', c: '' } // err
const c: IInput = { a: '', b: '', c: '' } // err
const d: IInput = { a: '' } // GOOD PASS
const e: IInput = { b: '' } // GOOD ERROR
const f: IInput = { b: '', c: ''  } // GOOD PASS
```
